src/trajectory.jl

```diff
@@ -115,90 +117,152 @@ Multinomial sampler carried during the building of the tree.
+    SliceTreeSampler(zcand, s.logu, (s.logu <= -energy(zcand)) ? 1 : 0)
 
 """
-Create a multinomial sampler for a single leaf tree:
-- the tree weight is just the probability of the only leave.
+Multinomial sampler for the starting single leaf tree.
+- tree weight is the (unnormalised) energy of the leaf.
 """
-makebase(s::MultinomialTreeSampler, H::AbstractFloat) = MultinomialTreeSampler(-H)
+MultinomialTreeSampler(s::MultinomialTreeSampler, H0::AbstractFloat, zcand::PhasePoint) = 
+    MultinomialTreeSampler(zcand, H0 - energy(zcand))
+
+function combine(rng::AbstractRNG, s1::SliceTreeSampler, s2::SliceTreeSampler)
+    @assert s1.logu == s2.logu "Cannot combine two slice sampler with different slice variable"
+    n = s1.n + s2.n
+    zcand = rand(rng) < s1.n / n ? s1.zcand : s2.zcand
+    SliceTreeSampler(zcand, s1.logu, n)
+end
+
+function combine(zcand::PhasePoint, s1::SliceTreeSampler, s2::SliceTreeSampler)
+    @assert s1.logu == s2.logu "Cannot combine two slice sampler with different slice variable"
+    n = s1.n + s2.n
+    SliceTreeSampler(zcand, s1.logu, n)
+end
+
+function combine(rng::AbstractRNG, s1::MultinomialTreeSampler, s2::MultinomialTreeSampler)
+    logw = logaddexp(s1.logw, s2.logw)
+    zcand = rand(rng) < exp(s1.logw - logw) ? s1.zcand : s2.zcand
+    MultinomialTreeSampler(zcand, logw)
+end
+
+function combine(zcand::PhasePoint, s1::MultinomialTreeSampler, s2::MultinomialTreeSampler)
+    logw = logaddexp(s1.logw, s2.logw)
+    MultinomialTreeSampler(zcand, logw)
+end
+
+mh_accept(
+    rng::AbstractRNG,
+    s::SliceTreeSampler,
+    s′::SliceTreeSampler
+) = rand(rng) < min(1, s′.n / s.n)
+
+mh_accept(
+    rng::AbstractRNG,
+    s::MultinomialTreeSampler,
+    s′::MultinomialTreeSampler
+) = rand(rng) < exp(min(0, s′.logw - s.logw))
+
+##
+## Variants of no-U-turn criteria
+##
 
-combine(s1::SliceTreeSampler, s2::SliceTreeSampler) = SliceTreeSampler(s1.logu, s1.n + s2.n)
-combine(s1::MultinomialTreeSampler, s2::MultinomialTreeSampler) = MultinomialTreeSampler(logaddexp(s1.logw, s2.logw))
+abstract type AbstractTerminationCriterion end
+
+struct NoUTurn <: AbstractTerminationCriterion end
+
+NoUTurn(::PhasePoint) = NoUTurn()
+
+struct GeneralisedNoUTurn{TR<:Real,TV<:AbstractVector{TR}} <: AbstractTerminationCriterion
+    rho::TV
+end
+
+GeneralisedNoUTurn(z::PhasePoint) = GeneralisedNoUTurn(z.r)
+
+combine(cleft::T, cright::T) where {T<:NoUTurn} = T()
+combine(cleft::T, cright::T) where {T<:GeneralisedNoUTurn} = T(cleft.rho + cright.rho)
+
+
+##
+## NUTS
+##
 
 """
 Dynamic trajectory HMC using the no-U-turn termination criteria algorithm.
 """
 struct NUTS{
+    S<:AbstractTreeSampler,
+    C<:AbstractTerminationCriterion,
     I<:AbstractIntegrator,
-    F<:AbstractFloat,
-    S<:AbstractTreeSampler
+    F<:AbstractFloat
 } <: DynamicTrajectory{I}
-    integrator  ::  I
-    max_depth   ::  Int
-    Δ_max       ::  F
-    samplerType ::  Type{S}
+    integrator      ::  I
+    max_depth       ::  Int
```

Comment:
Perhaps `Int` ==> `I`?

Reply:
What does this mean?

Reply:
oops, my fault - I thought `I` is a subtype of `Integer`.